### ISSUE TITLE
Interface to sample/ERPs

### ISSUE THREAD

**Issue by null-a** [probmods/webppl#246]
Thinking about some of the open issues relating to `ERP`s (#119, #231) has made me think about the choice to have the interface to `sample` be:

````js
sample(erp, params);
````

Rather than:

````js
sample(erp(params));
````

Having thought about it a little, it seems to me that switching to the second of these might have some benefits.

More specifically, the suggestion would be to switch to having a function represent a family of distributions (I'll call these `Dist`s here to disambiguate them from `ERP`s) which when called returns a particular distribution from the family. It would look like this:

````js
var coin = bernoulliDist(0.5);
coin.sample();
coin.score(true);
sample(coin);
````

Switching to this approach might help in the following ways:

* It might help to simplify the inference code which deals with importance distributions a little. For example at present `MHKernel` is concerned with [constructing arrays of parameters](https://github.com/probmods/webppl/blob/7e4ff8f3544c323c6c389454b139323e28fbb5e0/src/inference/mhkernel.js#L62) to hand off to drift kernels. This is already a little messy and when I spent a little time thinking about #231 it seemed its implementation would make things worse. The hope is that by bundling the `erp` and `params` together, working with `Dist`s would allow us to more easily factor this massaging of parameters out of the inference code.

* It would allow us to remove the current notion of [parameterized](https://github.com/probmods/webppl/blob/7e4ff8f3544c323c6c389454b139323e28fbb5e0/src/erp.js#L69) and unparameterized `ERP`s.

This, for example:

````js
var FairCoin = bernoulliERP.withParameters([0.5]);

var model = function() {
  var x = sample(FairCoin);
  var y = sample(withImportanceDist(FairCoin, bernoulliERP.withParameters([x ? .5 : 1])));
  return [x, y];
};
````

Would become something like:

````js
var FairCoin = bernoulliDist(0.5);

var model = function() {
  var x = sample(FairCoin);
  var y = sample(withImportanceDist(FairCoin, bernoulliDist(x ? .5 : 1)));
  return [x, y];
};

````

* We already have a few functions in `header.wppl` which behave rather like `Dist`s. For example, you might expect `deltaERP` to behave like this:

````js
sample(deltaERP, [0]) => 0
deltaERP.score([0], 0) => 0
deltaERP.score([0], 1) => -Infinity
````

However, `deltaERP` is actually a function which returns an "unparameterized" `ERP`, so it actually behaves like this:

````js
sample(deltaERP(0), []) => 0
deltaERP(0).score([], 0) => 0
deltaERP(0).score([], 1) => -Infinity
````

Removing this inconsistency seems like a reasonable thing to do. Under the proposed change this would become:

````js
sample(deltaERP(0)) => 0
deltaERP(0).score(0) => 0
deltaERP(0).score(1) => -Infinity
````

Now, it seems very likely that this will have already been considered at some point, so perhaps there are hidden pitfalls I'm not seeing. Performance is one possibility that comes to mind, although I think it might be possible to implement this with little overhead. Another is that some inference algorithms may somehow make use of easy access to the `params` array (`ERP` equality checks in `IncrementalMH` perhaps?) but that seems like something which `Dist`s could accommodate if necessary.

I'm not entirely sure this is the right way to go, but it seemed to be worth raising.

**Comment by dritchie** [comment]
Indeed, `IncrementalMH` does perform parameter equality checks, so ERPs of this form would need to provide some (efficient) equality check.

I agree that performance is a concern, since for ERPs with dynamically-determined parameters, this will lead to the creation of lots of ERP objects. Then again, with the current interface, ERP parameters are always packaged into `Array` objects before being passed to `sample`, so maybe this isn't such a big deal.

**Comment by stuhlmueller** [comment]
As far as I can tell, this proposal mirrors [how Anglican deals with distributions](http://www.robots.ox.ac.uk/~fwood/anglican/language/index.html#distributions). My sense is that this is a reasonable way of doing things if performance and equality check issues can be addressed, if we're sure that this will play nicely with all existing inference algorithms, and if the ERP-related definitions in `erp.js` look at least as good as under the current system. Of these, the potential performance hit is my biggest concern.

**Comment by null-a** [comment]
> with the current interface, ERP parameters are always packaged into Array objects before being passed to sample

Yeah, the idea that we might be able to loose that array is one of the reasons I think this is do-able.

Even if we don't, and the extra object adds overhead, I imagine it's a tiny constant factor that's hard to even notice.

> if we're sure that this will play nicely with all existing inference algorithms

One reason to think this *might* be the case is that I did things this way in [stochy](http://nbviewer.ipython.org/github/null-a/StochyExamples/blob/master/Marbles.ipynb) and didn't hit any problems. (In that example the `~` operator is a macro which expands to a call to `sample`.) Though I didn't implement `IncrementalMH` of course.

**Comment by null-a** [comment]
> if the ERP-related definitions in erp.js look at least as good as under the current system

Including the suggestion from #268, an implementation of this might look something like this:

````js
// erp.js

function bernoulliERP(s, k, a, params) {
  var dist = _.create(bernoulliProto, params);
  _.defaults(dist, bernoulliDefaults);
  return k(s, dist);
};

var bernoulliDefaults = { p: 0.5 };

var bernoulliProto = {
  sample: function() {
    return util.random() < this.p;
  },
  score: function(value) {
    if (value !== true && value !== false) {
      return -Infinity;
    }
    return value ? Math.log(this.p) : Math.log(1 - this.p);
  },
  support: function() {
    return [true, false];
  }
};

// header.wppl
var flip = function(p) {
  return sample(bernoulliERP({ p: p }));
};
````

I think this is fine, and in one respect I think it's a slight improvement over the [current approach](https://github.com/probmods/webppl/blob/c15f4affef9afb09309eb6def1404fc78f5ff4cf/src/erp.js#L143) as we don't have to duplicate `var weight = params[0];` in `sample` and `score` etc. (This also makes it easy to specify default parameters in one place which is nice.)

There would be a lot of repetition across the functions `bernoulliERP`, `gaussianERP`, etc. but we could perhaps address that using macros/meta-programming.

> Of these, the potential performance hit is my biggest concern.

I did quick initial test of the code above with a suitably modified version of MH, and drawing 10^6 samples from the model in `binomial.wppl` was no slower than before.

**Comment by null-a** [comment]
I've started work on this, and things appear to be working out so far. One question that has come up is what should `multiplexERP` look like under the new interface.

Looking at `examples/multiplex.wppl`, current use of this looks like:

````
var erpX = multiplexERP([true, false], [erp1, erp2]);
var toss = flip(.5);
sample(erpX, [toss]);
````

Under the new interface, `sample` only takes a single arg (the ERP) and the `params` are instead passed to the ERP constructor. So maybe we'd have:

````
var toss = flip(.5);
var erpX = multiplexERP([true, false], [erp1, erp2], toss);
sample(erpX);
````

But now it appears that the use of `multiplexERP` is overkill, as we can write something similar without it:

````
var toss = flip(.5);
var multi = { true: erp1, false: erp2 };
sample(multi[toss]);
````

(This isn't quite the same as `multiplexERP` because this will use `toString` to create the keys of `multi` rather than `util.serialize`, but you could imagine fixing that in other ways.)

With that as background, the questions I could use help with are:

- Is `multiplexERP` still useful under this interface?
- If not, is that a problem?
- If so, is the set-up in my second example OK?

**Comment by stuhlmueller** [comment]
Hmm, I thought I had replied to that comment a while ago. Apparently not. The gist of my hypothetical comment was that there is a specific use case where I needed to package up multiple ERPs into one, but that use case has gone away and could potentially have been addressed differently in the first place, so I'd get rid of `multiplexERP` and go with the final solution you presented.

**Comment by null-a** [comment]
Closed by #393.
